lib/event.c

```diff
@@ -848,12 +975,175 @@ static void thread_free(struct event_loop *master, struct event *thread)
 	XFREE(MTYPE_THREAD, thread);
 }
 
+#if EPOLL_ENABLED
+/*
+ * flush new event requests to epoll_fd set and m->handler.events
+ *
+ * @REQUIRE m->mtx
+ */
+static void update_events(struct event_loop *m)
+{
+	struct frr_epoll_event *new_frr_ev, *existing_frr_ev;
+	struct epoll_event ev, *new_ev, *existing_ev;
+	bool has_existing;
+	int update_fd;
+	struct stat fd_stat;
+
+	/* Flush m->handler.new_events to epoll_fd set and m->handler.events */
+	for (nfds_t i = 0; i < m->handler.new_event_count; i++) {
+		new_frr_ev = m->handler.new_event_ptrs[i];
+		new_ev = &new_frr_ev->ev;
+		update_fd = new_ev->data.fd;
+		existing_frr_ev = m->handler.events + update_fd;
+		existing_ev = &existing_frr_ev->ev;
+		has_existing = (existing_frr_ev->index >= 0);
+
+		if (fstat(update_fd, &fd_stat) == -1) {
+			zlog_debug("[!] In %s: fstat failed, update_fd: %d",
+				   __func__, update_fd);
+		}
+
+		/*
+		 * Construct ev which is passed to epoll_ctl call
+		 * Note that ev.events should be new events OR'd with
+		 * existing events
+		 */
+		ev.data.fd = update_fd;
+		ev.events = has_existing ? new_ev->events | existing_ev->events
+					 : new_ev->events;
+
+		/* Update epoll_fd set, .events and .event_ptrs */
+		if (!has_existing) {
+			if (S_ISREG(fd_stat.st_mode)) {
+				/*
+				 * Don't add regular file into epoll set,
+				 * directly add it into regular_revents.
+				 */
+				m->handler
+					.regular_revents
+						[m->handler
+							 .regular_revent_count]
+					.data.fd = update_fd;
+				m->handler
+					.regular_revents
+						[m->handler
+							 .regular_revent_count]
+					.events = ev.events;
+				m->handler.regular_revent_count++;
+			} else if (-1 == epoll_ctl(m->handler.epoll_fd,
+						   EPOLL_CTL_ADD, update_fd,
+						   &ev)) {
+				if (errno == EEXIST) {
+					/* We are never supposed to reach here
+					 */
+					zlog_debug(
+						"[!] Attempting to add an existing fd into epoll_fd set. Existing fd: %d, existing_ev: %d, req_ev: %d",
+						update_fd, existing_ev->events,
+						new_ev->events);
+					if (-1 == epoll_ctl(m->handler.epoll_fd,
+							    EPOLL_CTL_MOD,
+							    update_fd, &ev)) {
+						zlog_debug(
+							"[!] Both attemption to add into and modify an epoll_event in epoll_fd set both failed. Existing fd: %d, existing_ev: %d, req_ev: %d",
```

Comment:
I do not understand this comment.  What is `attemption`

Reply:
This if statement and zlog_debug is deleted. They are handled by zlog_err above.
It seems that I left these lines when I was debugging. Anyway they are useless now.